rfd/0074-sftp-support.md

```diff
@@ -0,0 +1,136 @@
+for per node.
+
+2. Adding SFTP protocol support to `tsh scp` sub command, and
+ensuring SFTP trasfers work on the web UI. `tsh scp` will continue to use the scp/rcp
+protocol by default for backwards compatibility, but the SFTP protocol can be
+optionally enabled with the `-s` flag.
+
+3. Making the SFTP protocol be used by default for
+`tsh scp`, and adding the `-O` flag to allow optional usage of thee scp/rcp protocol
+for backwards compatibility. This stage will take place in a future major release
+(11?) and will be documented as a potentially breaking change.
+
+Both the `-s` and `-O` flags are what OpenSSH uses to allow the user to choose
+what protocol to use, and the deprecation process of making SFTP optional then
+default is what OpenSSH did as well.
+
+A potential fourth stage would be adding the `tsh sftp` sub command which would
+be very similar in behavior to OpenSSH's `sftp` command. This will only be done
+if users express a need for it.
+
+#### Roles
+
+A role option `disable_file_copy` will be added that will define if file
+copying via scp/rcp or sftp protocols will be allowed. This role option will
+be false by default.
+
+If a user has multiple roles that have different values for `disable_file_copy`,
+then file copying will be disabled if the role restricting file copying matches
+the server the user is trying to access. For example:
+
+```yaml
+ind: role
+version: v5
+metadata:
+  name: allow-copy-test
+options:
+  disable_file_copy: true
+spec:
+  allow:
+    labels:
+      env: test
+
+kind: role
+version: v5
+metadata:
+  name: deny-copy-prod
+options:
+  disable_file_copy: false
+spec:
+  allow:
+    labels:
+      env: prod
+```
+
+File copying would be disabled for nodes with label "env: prod" but enabled
+for "env: test" if the user has both of these roles attached.
+
+### Security
+
+SFTP uses the SSH protocol to provide confidentiality and integrity, as it is
+used inside an existing SSH connection. It also has the benefit of not starting
+processes with attacker-controlled arguments like the scp/rcp protocol does.
+
+As mentioned above, Teleport Node services will re-execute themselves as the
+SSH login user to handle SFTP connections. This will ensure users can only
+access and modify files they are allowed to.
+
+### UX
+
+The UX of `tsh scp` with SFTP enabled will be very similar to that of `tsh scp`
+with scp/rcp enabled, but there will be some differences of behavior between the
+two. SFTP does not natively support expanding home directories, so absolute paths
+will have to be used for remote paths. Likewise passing quoted commands to be 
+remotely executed (ex. ``tsh scp my_file 'user@host:/tmp/`whoami`.txt'``) will
+not work as `teleport sftp` is run without any other arguments.
+
+The web UI for trasferring files will need to have a method of allowing the
+user to choose between scp and SFTP. A knob could be added for this purpose,
```

Comment:
Let's remove the knob. The UI should automatically switch to SFTP when it's available until major upgrade. In major upgrade the ui should always use SFTP.

Reply:
What if a user wants to transfer files to an older OpenSSH server from the web UI though? After the major version they couldn't without upgrading their OpenSSH server

Reply:
they can always use CLI for that?

Reply:
True, I guess the goal is to keep the web UI simple